### DISCUSSION TITLE
出题入门指北

### DISCUSSION THREAD

**Discussion by DOGGOD-Q** [CSGrandeur/s-1problem1day1ac#119] [Knowledge]
# 出题入门指北

## 写在前面

很多 acmer 都有一个出题的梦想。幽默风趣的题面小故事、整洁漂亮的数学公式、恰到好处的时空范围和数据大小、干净优雅的解题思路以及难以揣测的极限数据，以上这些东西等你看完的本篇文章后——你还是不一定能做到。但是这篇文章足以带你入门，帮助你完成人生的前几次自主命题。等到熟悉了这一切，你所追求的东西也将水到渠成。

本文由2020深圳技术大学第一届ACM程序设计迎新赛命题组成员联合提供，请在环境明亮处阅读

## 出题Markdown使用教程补充

前置学习条件：（https://github.com/CSGrandeur/s-1problem1day1ac/discussions/116）

###  一、推荐的编辑器

1. Typora （强烈推荐）

   - **免费** 、**实时预览**、**扩展语法**、**跨平台**、**简洁美观**

2. VS code

#### 下载地址

- （https://www.typora.io/）
- 下滑找到右上角的Download字样完成下载

#### 界面介绍

![微信图片_20210304182839](https://user-images.githubusercontent.com/35170890/109957372-cfe0b380-7d1f-11eb-9c5d-cf5d991906e8.png)

- 左边文件可以随时查看文件，但需要先保存正在编辑的文件，否则会造成文件丢失，Typora会有提示的，请放心。
- 大纲可以随时查看当前编写文件的树型结构。
- 其他的大家自行探索，不难的。

#### 主要功能介绍

在一开始发的前置学习条件中（https://github.com/CSGrandeur/s-1problem1day1ac/discussions/116）

### 三、出题中Markdown

1. oj支持 mathjax ，而 Typora 也支持 mathjax（但是 github 不支持…） ，右键选择`“插入”->“代码块”`,或者连续输入两个`“$$”`后按回车即可调出mathjax ,必须是单独的一行,例如：

$$
f[i] = f[i - 1] + f[i - 2]
$$

2. 在Typora中换行仅需要回车

3. 分段：按正常格式书写，Typora是自动补全分段

4. 特殊字体

   - 在需要 **加粗** 的文字前后分别加上2个“ ** ”可以完成。`**（就像这样）**`

   - 在需要 ~~删除线~~ 的文字前后分别加上2个“ ~~ ”可以完成。`~~（就像这样）~~`

   - 在需要 *斜体*  的文字前后分别加上1个“ * ”可以完成。`*（就像这样）*`

   - 在需要 ***斜体加粗***  的文字前后分别加上3个“ * ”可以完成。`***（就像这样）***`

   - 在需要 ***~~斜体加粗删除线~~***  的文字前后对称加上“ ~ ~ * * * ”可以完成。`***~~（就像这样）~~***`
   - 剩下的大家自由组合一下可以出来

5. 在Typora中Markdown支持插入图片，但图片都是以本地的形式保存，如果直接拖拽至oj中编辑框中，就会造成图片丢失；我们需要将图片先备份一份在本地，之后再将图片拖进oj中编辑框的里的相应位置就行，上传图片要一会时间，约为5秒左右

6. 依托于Typora的强大工能，在左下角可以启动源代码模式，之后选择经典操作`“Ctrl  + C”`再复制到oj上的文本框中就好了。

### 题目背景

针对ACM的出题，我们要在题目中清晰的给出所有题目的信息，题目背景可以有，但是题目背景不能影响题意的理解，当背景过长时，把题目考察的问题放到故事结束比较合适，不要强行融合进故事中，把题目变成做阅读理解。

### 题目描述

在涉及题意的关键语句中，我们没必要节省篇幅压缩到一个句子中，应该把这个事情用更详细的方式描述出来，下面是一个例子。

```
求s到t的路径中所有边权的最大值的最小值
```

如果我们拆分一下，它可以变成这样

```
定义一条从 s到t的路径中所有边的边权最大值为： s到t的 “瓶颈值”. 求s到t的最小瓶颈值
```

虽然题意没有变化，在做题的时候这样无疑可以得到更好的做题体验。不管是英文题目还是中文题目我们不能未定义直接使用一个概念，如图论题中，你要用传送门来代表两点之间的路径，你需要先解释传送门的意义。放置不被理解。并且题目中每个概念都要用同一个词语来表示，比如，不应该一会说费用，一会说代价。不要在题目中写一些影响理解的话，保证题面格式清晰。

### 关于公式

一些数学题中，我们要用到一些数学公式，对于这些公式，我们最好使用LaTeX来进行编写，这里给大家一个方便的LaTeX编辑网站

https://www.codecogs.com/latex/eqneditor.php

### 输入输出格式

在写输入输出的时候，我们要尽可能简短地把输入的值和题目中的变量对应起来，介绍明白每个变量的意义，题目的数据范围要在输入输出中表示清楚，如果题目出现了小数，涉及到精度问题，需要出题人自己写一个特判代码来防止轻度误差。

下面给出几个简单的例子

![QQ图片20210304201330](https://user-images.githubusercontent.com/35170890/109962719-7f208900-7d26-11eb-9329-2880a081b4d9.png)

### 题目时间复杂度和标程

出题人在确定了题目考察的算法后，把自己的idea转化成一个问题，从各种角度出发，考虑题目的最佳复杂度，然后指定合适的数据范围，编写标准程序。标程确保可以通过所有极限数据和边缘数据

再设置题目时间。为了保证能卡掉复杂度错误的暴力代码，我们的极限数据要指定合理。评测机1s大概可以处理1e8的数据，可以根据这个数据来推算各种复杂度的极限数据。在这些工作全部结束后，我们需要保证我们的边界不能误伤相同复杂度的代码，也就是不能卡常。通常来讲，我们可以让自己的代码跑极限数据进行尝试，如果1s的时间很极限卡过，那么题目时间可以给到2s，保证标程2倍以内的所有代码都能跑过，需要强调的是，标程不能用一些奇怪的方式来进行优化，一切操作都选用正常操作，比如读入不要用快读。

### 样例和样例解释

样例应当有一定的强度，能够查出一些简单的错误。读错题意的人应当能够通过样例发现自己读错了题意。

有多种操作的题，每种操作都应在样例中出现。

有多种输出的题，每种输出都应在样例中出现。例外：实际上不可能无解，但要求判断是否有解的题目。

题目描述越复杂、越不易理解就越应当有详细的样例解释。

详细的样例解释可以选择配上图片。

较大的样例可以没有样例解释。

如果题目的样例给出直接影响做答，也可以不给出样例解释，或只给出一个很小样例的解释。

## 制作数据
出题时我们经常需要批量生成数据，可以随便用C/C++写个Generator。

### Generator组成
一般可分为正解程序和生成器两部分，生成器用来生成输入，正解程序用来生成输出。

#### 正解程序

此处正解程序需要进行小修改，如原正解程序：

```cpp
#include <cstdio>

int a, b;

int main() {
	scanf("%d%d", &a, &b);
	printf("%d\n", a + b);
	return 0;
}
```

我们进行如下修改：

```cpp
#include <cstdio>

int a, b;

int main(int argc, char* argv[]) {
	freopen(argv[1], "r", stdin);
	freopen(argv[2], "w", stdout);

	scanf("%d%d", &a, &b);
	printf("%d\n", a + b);
	return 0;
}
```

我们可以看见主要是 main 函数参数变化和 freopen 的添加。

main 函数的 argc 参数代表程序调用时的参数数量，`argv[]` 则是参数字符串。

如在命令行调用：

`test.exe asd sdf dfg`

则 `test.exe` 可以得到 `argc = 4, argv[] = {test.exe 的绝对路径, "asd", "sdf", "dfg"}`

在无调用参数的情况下，`argc = 1, argv[] = {test.exe 的绝对路径}`

这样修改正解程序是为了方便生成器调用其生成输出文件

freopen 则是使用文件输入输出代替命令行读数/写入数据

#### 生成器
生成器样例：

```cpp
#include <cstdio>
#include <cstdlib>

#include <string>

// 生成数据组数
#define NUM 10
// 正解程序名
#define SLN "solution.exe"

int main() {
	for (int T = 0; T < NUM; ++T) {
		// 生成输入or输出文件名
		std::string iname = "t", oname = "t";
		iname += std::to_string(T) + ".in";
		oname += std::to_string(T) + ".out";

		// 链接写入输入文件
		freopen(iname.c_str(), "w", stdout);

		/*
			这里填入你的数据生成程序，直接用标准输出即可
		*/

		// 关闭文件
		fclose(stdout);

		// 生成调用命令
		std::string call(SLN);
		call += ' ';
		call += iname + ' ' + oname;

		// 调用正解程序
		system(call.c_str());
	}
	return 0;
}
```

通过这个最简单的模板就可以批量生成数据了。

请确保正解程序与生成器在同一目录下。

### 数据内容

一道题的数据通常包含以下几种。

完全的随机数据：代表最一般的情况，但这并不意味着它一定是最水的数据；

边界数据：这里主要是考验做题者思维的严谨性，比如说答案恰好在搜索的起点、整个图只有一个点、质数相关题目中的数字 1 、动态规划的边界情况、棋盘的边缘等等；

专门针对某些算法的极限数据：靠特殊情况使某些较容易的算法退化，比如 SPFA 与“菊花图”、快速排序与倒序数等等。

## 对拍

你现在有了标程，有了数据，为了不出问题你还写了几种可能的错误程序，那么应该怎样测试你的数据能不能卡住这些错误程序呢？虽然某些情况下我们可以直接用肉眼检查，但是如果输出数据太长或者想要测试的数据太多，这时再用眼睛去看就显得有点蠢了：这种机械化的任务就应该让电脑来完成。

这时就需要**对拍程序**出场了。对拍程序实现的功能很简单：

    1.生成一组数据
    2.分别运行标程和待测试程序
    3.对比输出结果，相同则回到第一步，否则暂停

这种功能我们可以利用**批处理指令**完成。以下举一个小例子

题目：输入一个整数 `n(1 <= n <= 10000)` 以及 n 个整数，请将这 n 个数从小到大排序后输出。

标程：`std.cpp`
```cpp
#include <cstdio>
#include <queue>
using namespace std;

int main()
{
    int n;
    scanf("%d", &n);
    priority_queue<int, vector<int>, greater<int>> h;
    for(int i=1; i<=n; i++)
    {
        int num;
        scanf("%d", &num);
        h.push(num);
    }
    for(int i=1; i<=n; i++)
    {
        printf("%d ", h.top());
        h.pop();
    }
    printf("\n");
    return 0;
}
```

待测试程序：`code.cpp`
```cpp
#include <cstdio>
#include <algorithm>
using namespace std;

int a[11010];

int main()
{
    int n;
    scanf("%d", &n);
    for(int i=1; i<=n; i++)
    {
        scanf("%d", &a[i]);
    }
    sort(a+1, a+n+1);
    for(int i=1; i<=n; i++)
    {
        printf("%d ", a[i]);
    }
    printf("\n");
    return 0;
}
```

随机数据生成程序：`datamaker.cpp`
```cpp
#include <cstdio>
#include <ctime>
#include <cstdlib>

using namespace std;

int main()
{
    srand(time(NULL));

    int n = (rand() * rand()) % 10000;
    printf("%d\n", n);
    for(int i=1; i<=n; i++)
    {
        printf("%d ", rand());
    }
    printf("\n");
    return 0;
}
```

将三个程序编译好后放到同一个文件夹中

![文件夹](https://user-images.githubusercontent.com/35170890/109957438-e5ee7400-7d1f-11eb-86be-50656300076c.png)

在文件夹中新建一个文件 `duipai.bat` ，右键选择编辑，输入以下代码

```
@echo off
:loop
    datamaker.exe > data.in
    std.exe < data.in > std.out
    code.exe < data.in > code.out
    fc std.out code.out
    if not errorlevel 1 goto loop
pause
goto loop
```

这些就是批处理指令。`@echo off`的意思是关闭指令显示，这样运行后的命令行中会更清晰（可以删掉试一下），`:loop` 和两个 `goto loop` 组成了一个循环结构；`std.exe < data.in` 的意思是将 `data.in` 作为 `std.exe` 的输入，`datamaker.exe > data.in` 的意思是将 `datamaker.exe` 的输出写入 `data.in`；`fc std.out code.out` 意为对比两个文件的文本是否相同，如果不同则将 errorlevel 的值变为 1 。整个代码还是很简明的。运行效果如下

![运行](https://user-images.githubusercontent.com/35170890/109957464-eedf4580-7d1f-11eb-839e-132d73e81d10.png)

这么厉害的对拍程序如果只用来出题就太可惜了，其实对拍在赛场上也能发挥巨大的作用。比如你写出了一个复杂度优秀但是无法证明正确性的程序，你就可以将这个程序与一个一定正确的暴力程序在小数据范围内对拍，如果拍了很久也没出现问题，那么这个程序大概率也不会有什么问题。

## 大功告成？

完成了以上出题的主要步骤之后，我们还要将题目上传到 OJ 上。每个 OJ 都有自己的特点，所以一定要仔细测试。一个比较常见的问题是不同系统的换行符并不相同（windows 是`\r\n`，Unix 是`\n`，Mac 是`\r`），而 c/cpp 程序使用 scanf 读入时是按照 Unix 的方式读入的，因此读到 windows 中**手写**的换行符时会发生意想不到的错误。诸如此类的问题只能靠严谨敏锐的观察和丰富的经验来避免，因此一定要仔细仔细再仔细。

你的程序正式投入比赛后，你要时刻关注参赛者的完成情况，看看大家的情况是否都在你的意料之中。如果发现意外，请坦荡承担一切，认真负责地解决大家的问题。

比赛结束，尘埃落定。现在你可以骄傲地发出自己的题解和标程，参与大家对题目的讨论，找个好机会把自己的出题历程拿来跟好朋友吹一吹，一起哈哈大笑之后继续向前吧。

**Comment by CSGrandeur** [comment]
在windows下制作题目数据时，所有 freopen的"w"、"r" 推荐改为 "wb"、"rb"，可躲过OJ的大多玄学问题。

**Comment by CSGrandeur** [comment]
多文件数据生成模板

以生成 输入一个整数、输出相同的整数 为例。

gen.cpp

```cpp
#include<stdio.h>
#include<string.h>
#include<stdlib.h>
char bufin[20], bufout[20];
int main()
{
    for(int i = 0; i < 20; i ++)
    {
        sprintf(bufin, "test%02d.in", i);
        FILE *fin = fopen(bufin, "wb");
        fprintf(fin, "%d\n", i);
        fclose(fin);
    }
    return 0;

}
```

solution.cpp

```cpp
#include<stdio.h>
#include<string.h>
#include<stdlib.h>
int main()
{
// #ifdef 与 #endif 圈起来的代码，只有在编译带有后面的参数（这里是“CPC”）时才会执行
// g++ -D CPC main.cpp -o main
// OJ上没有这个编译参数，便不会执行这部分代码，从而本地可生成数据，交到OJ又可以正常AC
#ifdef CPC
char bufin[20], bufout[20];
for(int i = 0; i < 20; i ++)
{
    sprintf(bufin, "test%02d.in", i);
    sprintf(bufout, "test%02d.out", i);
    freopen(bufin, "rb", stdin);
    freopen(bufout, "wb", stdout);
#endif

    int a;
    while(scanf("%d", &a) != EOF)
    {
        printf("%d\n", a);
    }

#ifdef CPC
}
#endif
    return 0;
}
```

**Reply by CSGrandeur** [reply]
随机数据，建议 `srand()` 传入一个固定参数当种子，从而每次执行得到相同随机数。比如 `srand(2021)`

`rand()` 随机一个整数，有丰富的用法，比如可以 

- `rand` 相除： `(double)rand() / (rand() + 1)`
- 也可以二进制拼接: `(unsigned long long)rand() << 32 | rand()`
- 变字符串拼接： `sprintf(buf, "%s%s", rand(), rand())`
- 随机字母： `rand() % 26 + 'A'`
- 处理负数： `int a = rand() & 1 ? 1 : -1;`
- 概率：`if(rand() % 100 > 80)`

**Reply by liu-ethan** [reply]
老师，请问我现在有solution.cpp和gen.cpp文件，
然后我用g++ -D CPC solution.cpp -o solution运行了solution.cpp文件
然后我用g++ gen.cpp -o gen运行了gen.cpp文件
文件是按照你上面的代码格式写的，但现在只生成了t0.in~t9.in，请问怎么生成.out文件
包括你给的大作业模板也是，按照上面那两行命令，只能生成in不能生成out

**Reply by DOGGOD-Q** [reply]
gen.cpp 是数据生成器，用于生成 .in 文件；solution.cpp 是标准程序，用于根据 .in 文件生成 .out 文件
所以要先运行 gen 再运行 solution

**Comment by CSGrandeur** [comment]
## Special Judge

spj 用于答案不唯一或精度有浮动的题目，通过一个程序，根据用户程序的输出和题目的答案输出做一定的逻辑判断，返回判题结果。

比如题目输出浮点误差不超过 `1e-6`，就可以用程序判断 `if(abs(user_out - sys_out) > 1e-6) return WA;'

比如拓扑排序答案不唯一，就可以用 spj 程序判断用户输出的结果是否合法。

```cpp
#define AC 0
#define WA 1

#include<stdio.h>
#include<math.h>
const double eps = 2e-3;
char dummy[100000];
int main(int argc, char **args){
    FILE * f_in = fopen(args[1], "r"); // 标准输入
    FILE * f_out = fopen(args[2], "r"); // 标准输出
    FILE * f_user = fopen(args[3], "r"); // 用户输出
    
    double oSys, oUser;
    while(fscanf(f_out, "%lf", &oSys) != EOF)
    {
    	if(fscanf(f_user,"%lf",&oUser) != 1) return WA;
    	if(abs(oSys - oUser) > eps) return WA;
    }
    if(fscanf(f_user,"%s",dummy)==1) return WA;
    fclose(f_in);
    fclose(f_out);
    fclose(f_user);
    return AC;
}
```

将`spj.cc`（必须用这个文件名）和测试数据一起上传，在OJ上对应题目启用 special judge，spj就生效了。
